File: android/src/main/java/app/edge/reactnative/core/EdgeNative.java
package app.edge.reactnative.core;

import android.util.Base64;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Iterator;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

class EdgeNative {
  private final Disklet mDisklet;
  private final ExecutorService mPool = Executors.newCachedThreadPool();

  public EdgeNative(File base) {
    mDisklet = new Disklet(base);
  }

  /**
   * Handles a native method call on a separate worker thread.
   *
   * <p>The promise resolution will also happen on this worker thread, so be prepared to bounce back
   * to the UI thread if necessary.
   */
  public void call(@NonNull String name, @NonNull String args, @NonNull PendingCall promise) {
    mPool.execute(
        new Runnable() {
          @Override
          public void run() {
            try {
              handleCall(name, new JSONArray(args), promise);
            } catch (Throwable error) {
              promise.reject(error.getMessage());
            }
          }
        });
  }

  private void handleCall(
      @NonNull String name, @NonNull JSONArray args, @NonNull PendingCall promise)
      throws IOException, JSONException {
    switch (name) {
      case "diskletDelete":
        mDisklet.delete(args.getString(0));
        promise.resolve(null);
        break;

      case "diskletGetData":
        promise.resolve(Base64.encodeToString(mDisklet.getData(args.getString(0)), Base64.NO_WRAP));
        break;

      case "diskletGetText":
        promise.resolve(mDisklet.getText(args.getString(0)));
        break;

      case "diskletList":
        promise.resolve(new JSONObject(mDisklet.list(args.getString(0))));
        break;

      case "diskletSetData":
        mDisklet.setData(args.getString(0), Base64.decode(args.getString(1), Base64.DEFAULT));
        promise.resolve(null);
        break;

      case "diskletSetText":
        mDisklet.setText(args.getString(0), args.getString(1));
        promise.resolve(null);
        break;

      case "fetch":
        handleFetch(args, promise);
        break;

      case "randomBytes":
        {
          SecureRandom sr = new SecureRandom();
          byte[] entropy = new byte[args.getInt(0)];
          sr.nextBytes(entropy);
          promise.resolve(Base64.encodeToString(entropy, Base64.NO_WRAP));
        }
        break;

      case "scrypt":
        {
          byte[] data = Base64.decode(args.getString(0), Base64.DEFAULT);
          byte[] salt = Base64.decode(args.getString(1), Base64.DEFAULT);
          int n = args.getInt(2);
          int r = args.getInt(3);
          int p = args.getInt(4);
          int dklen = args.getInt(5);
          byte[] out = scrypt(data, salt, n, r, p, dklen);
          if (out == null) promise.reject("Failed scrypt");
          else promise.resolve(Base64.encodeToString(out, Base64.NO_WRAP));
        }
        break;

      default:
        promise.reject("No method " + name);
    }
  }

  private void handleFetch(@NonNull JSONArray args, @NonNull PendingCall promise)
      throws JSONException {
    String uri = args.getString(0);
    String method = args.getString(1);
    JSONObject headers = args.getJSONObject(2);
    String body = args.optString(3);
    boolean bodyIsBase64 = args.optBoolean(4);

    HttpURLConnection connection = null;
    try {
      // Set up the HTTP connection:
      connection = (HttpURLConnection) new URL(uri).openConnection();
      connection.setRequestMethod(method);
      connection.setDoInput(true);
      connection.setUseCaches(false);

      // HTTPS needs help:
      if (connection instanceof HttpsURLConnection) {
        HttpsURLConnection httpsConnection = (HttpsURLConnection) connection;
        SSLContext context = SSLContext.getInstance("TLS");
        context.init(null, null, null);
        httpsConnection.setSSLSocketFactory(context.getSocketFactory());
      }

      // Add the headers:
      for (Iterator<String> it = headers.keys(); it.hasNext(); ) {
        String key = it.next();
        connection.setRequestProperty(key, headers.getString(key));
      }

      // Add the body:
      if (body.length() > 0) {
        byte[] bodyData =
            bodyIsBase64
                ? Base64.decode(body, Base64.DEFAULT)
                : body.getBytes(StandardCharsets.UTF_8);
        connection.setRequestProperty("Content-Length", Integer.toString(bodyData.length));
        connection.setDoOutput(true);
        OutputStream outStream = connection.getOutputStream();
        outStream.write(bodyData);
        outStream.flush();
        outStream.close();
      }

      // Make the request:
      connection.connect();

      // Read the response status:
      JSONObject response = new JSONObject();
      response.put("status", connection.getResponseCode());

      // Read the response headers:
      JSONObject responseHeaders = new JSONObject();
      for (int i = 0; true; ++i) {
        @Nullable String key = connection.getHeaderFieldKey(i);
        @Nullable String value = connection.getHeaderField(i);
        if (key == null && i == 0) continue; // Might be the HTTP status
        if (key == null || value == null) break;
        responseHeaders.put(key, value);
      }
      response.put("headers", responseHeaders);

      // Read the response body:
      StreamStringReader responseBody = new StreamStringReader();
      responseBody.read(connection.getInputStream(), responseHeaders.optInt("Content-Length"));
      try {
        response.put("body", responseBody.getUtf8());
        response.put("bodyIsBase64", false);
      } catch (Exception error) {
        response.put("body", responseBody.getBase64());
        response.put("bodyIsBase64", true);
      }

      promise.resolve(response);
    } catch (Exception error) {
      promise.reject("Native fetch: " + error.getMessage());
    } finally {
      if (connection != null) connection.disconnect();
    }
  }

  private native byte[] scrypt(byte[] data, byte[] salt, int n, int r, int p, int dklen);

  static {
    System.loadLibrary("edge-core-jni");
  }
}

File: ios/EdgeNative.swift
class EdgeNative {
  let disklet = Disklet()
  let queue = DispatchQueue(label: "app.edge.reactnative.core")

  /**
   * Handles a native method call on a separate worker thread.
   *
   * The promise resolution will also happen on this worker thread,
   * so be prepared to bounce back to the UI thread if necessary.
   */
  public func call(
    _ name: String,
    args: NSArray,
    promise: PendingCall
  ) {
    return queue.async {
      do {
        try self.handleCall(name, args: args, promise: promise)
      } catch {
        promise.reject(String(describing: error))
      }
    }
  }

  func handleCall(
    _ name: String,
    args: NSArray,
    promise: PendingCall
  ) throws {
    if name == "diskletDelete", let path = args[0] as? String {
      try disklet.delete(path: path)
      return promise.resolve(nil)
    }

    if name == "diskletGetData", let path = args[0] as? String {
      return promise.resolve(try disklet.getData(path: path).base64EncodedString())
    }

    if name == "diskletGetText", let path = args[0] as? String {
      return promise.resolve(try disklet.getText(path: path))
    }

    if name == "diskletList", let path = args[0] as? String {
      return promise.resolve(try disklet.list(path: path) as NSDictionary)
    }

    if name == "diskletSetData",
      let path = args[0] as? String,
      let base64 = args[1] as? String,
      let data = Data.init(base64Encoded: base64)
    {
      try disklet.setData(path: path, data: data)
      return promise.resolve(nil)
    }

    if name == "diskletSetText",
      let path = args[0] as? String,
      let text = args[1] as? String
    {
      try disklet.setText(path: path, text: text)
      return promise.resolve(nil)
    }

    if name == "fetch",
      let uri = args[0] as? String,
      let method = args[1] as? String,
      let headers = args[2] as? NSDictionary
    {
      let body = args[3]
      let bodyIsBase64 = args[4] as? Bool
      return handleFetch(
        uri: uri,
        method: method,
        headers: headers,
        body: body,
        bodyIsBase64: bodyIsBase64 ?? false,
        promise: promise)
    }

    if name == "randomBytes", let size = args[0] as? Int {
      if let entropy = NSMutableData(length: size),
        SecRandomCopyBytes(kSecRandomDefault, size, entropy.mutableBytes) == errSecSuccess
      {
        return promise.resolve(entropy.base64EncodedString())
      }
      return promise.reject("Could not obtain secure entropy")
    }

    if name == "scrypt",
      let data64 = args[0] as? String,
      let salt64 = args[1] as? String,
      let n = args[2] as? UInt64,
      let r = args[3] as? UInt32,
      let p = args[4] as? UInt32,
      let dklen = args[5] as? Int,
      let data = NSData.init(base64Encoded: data64),
      let salt = NSData.init(base64Encoded: salt64),
      let out = NSMutableData(length: dklen)
    {
      if crypto_scrypt(
        data.bytes.bindMemory(to: UInt8.self, capacity: data.length), data.length,
        salt.bytes.bindMemory(to: UInt8.self, capacity: salt.length), salt.length,
        n, r, p,
        out.mutableBytes.bindMemory(to: UInt8.self, capacity: dklen), dklen
      ) != 0 {
        return promise.reject("Failed scrypt")
      }
      return promise.resolve(out.base64EncodedString())
    }

    if name == "sqlOpen",
      let database = args[0] as? String,
      let key64 = args[1] as? String,
      let key = NSData.init(base64Encoded: key64)
    {
      var error: UnsafeMutablePointer<CChar>?
      let handle = edgeSqlOpen(
        databasePath(database),
        key.bytes.bindMemory(to: UInt8.self, capacity: key.length),
        Int32(key.length),
        &error)
      if handle < 0 {
        return promise.reject(takeError(&error, "Cannot open the database"))
      }
      return promise.resolve(Int(handle))
    }

    if name == "sqlExec",
      let handle = args[0] as? Int,
      let statements = args[1] as? String
    {
      var error: UnsafeMutablePointer<CChar>?
      return resolveJson(
        edgeSqlExec(Int32(handle), statements, &error), &error, promise)
    }

    if name == "sqlBatch",
      let handle = args[0] as? Int,
      let statements = args[1] as? String
    {
      var error: UnsafeMutablePointer<CChar>?
      return resolveJson(
        edgeSqlBatch(Int32(handle), statements, &error), &error, promise)
    }

    if name == "sqlQuery",
      let handle = args[0] as? Int,
      let sql = args[1] as? String
    {
      let params = args[2] as? String
      var error: UnsafeMutablePointer<CChar>?
      return resolveJson(
        edgeSqlQuery(Int32(handle), sql, params, &error), &error, promise)
    }

    if name == "sqlClose", let handle = args[0] as? Int {
      edgeSqlClose(Int32(handle))
      return promise.resolve(nil)
    }

    if name == "sqlDelete", let database = args[0] as? String {
      var error: UnsafeMutablePointer<CChar>?
      if edgeSqlDelete(databasePath(database), &error) != 0 {
        return promise.reject(takeError(&error, "Cannot delete the database"))
      }
      return promise.resolve(nil)
    }

    return promise.reject("No method \(name)")
  }

  /**
   * Where an account's database lives.
   *
   * Beside the disklet's own storage, so an account's database sits with the
   * rest of its device-local state and is removed with it.
   */
  func databasePath(_ name: String) -> String {
    let paths = NSSearchPathForDirectoriesInDomains(
      .documentDirectory, .userDomainMask, true)
    let base = URL(fileURLWithPath: paths[0]).appendingPathComponent("databases")
    try? FileManager.default.createDirectory(
      at: base, withIntermediateDirectories: true)
    return base.appendingPathComponent("\(name).db").path
  }

  /** Takes ownership of a native error string, falling back to `fallback`. */
  func takeError(
    _ error: inout UnsafeMutablePointer<CChar>?, _ fallback: String
  ) -> String {
    guard let error = error else { return fallback }
    let message = String(cString: error)
    edgeSqlFree(error)
    return message.isEmpty ? fallback : message
  }

  /** Resolves with a native JSON string, or rejects with its error. */
  func resolveJson(
    _ result: UnsafeMutablePointer<CChar>?,
    _ error: inout UnsafeMutablePointer<CChar>?,
    _ promise: PendingCall
  ) {
    guard let result = result else {
      return promise.reject(takeError(&error, "SQL failed"))
    }
    let json = String(cString: result)
    edgeSqlFree(result)
    return promise.resolve(json)
  }

  func handleFetch(
    uri: String,
    method: String,
    headers: NSDictionary,
    body: Any,
    bodyIsBase64: Bool,
    promise: PendingCall
  ) {
    // Set up the HTTP connection:
    guard let url = URL(string: uri) else {
      return promise.reject("Invalid URL")
    }
    var request = URLRequest(url: url)
    request.httpMethod = method

    // Add the headers:
    for (key, value) in headers {
      if let keyString = key as? String,
        let valueString = value as? String
      {
        request.setValue(valueString, forHTTPHeaderField: keyString)
      }
    }

    // Add the body:
    if let bodyText = body as? String {
      request.httpBody =
        bodyIsBase64
        ? Data.init(base64Encoded: bodyText)
        : bodyText.data(using: .utf8)
    }

    // Set up the response callback:
    let task = URLSession.shared.dataTask(with: request) { data, response, error in
      if let error = error {
        return promise.reject("Native fetch: \(error)")
      }

      guard let httpResponse = response as? HTTPURLResponse else {
        return promise.reject("Native fetch: Missing HTTPURLResponse")
      }
      let out = NSMutableDictionary()
      out["status"] = httpResponse.statusCode

      // Read the response headers:
      let headers = NSMutableDictionary()
      for (key, value) in httpResponse.allHeaderFields {
        if let keyString = key as? String,
          let valueString = value as? String
        {
          headers[keyString] = value
        }
      }
      out["headers"] = headers

      // Read the response body:
      if let bodyData = data {
        if let body = String(bytes: bodyData, encoding: .utf8) {
          out["body"] = body
          out["bodyIsBase64"] = false
        } else {
          out["body"] = bodyData.base64EncodedString()
          out["bodyIsBase64"] = true
        }
      }

      promise.resolve(out)
    }

    // Make the request:
    task.resume()
  }
}
